Add objects in java tasks

tsbonev opened the pull request.

mgenov commented on src/com/clouway/crm/core/Cars/Mercedes.java:
```diff
@@ -0,0 +1,13 @@
+
+import java.time.LocalDateTime;
+
+public class Mercedes extends Automobile {
```
The goal of this task is to be draw on a list of paper and class diagrams to be used.

mgenov commented on src/com/clouway/crm/core/Sumator/SumatorExceptions.java:
```diff
@@ -0,0 +1,46 @@
+import java.math.BigInteger;
+import java.text.ParseException;
+
+public class SumatorExceptions {
```
It should be just Sumator as only the exceptional class could end with exception by convention.

mgenov commented on src/com/clouway/crm/core/Sumator/SumatorExceptions.java:
```diff
@@ -0,0 +1,46 @@
+            }
+            */
+        }
+        finally {
```
Finally could be removed and success flag too. You may return and in the catch clause too.

mgenov commented on src/com/clouway/crm/core/Tree/HetTreeElement.java:
```diff
@@ -0,0 +1,32 @@
+
+public class HetTreeElement<T> {
+
+    static int elemCount = 0;
```
Doesn't look good as it's a global variable and will count elements globally for the application. They should be used in really rare cases.

mgenov commented on src/com/clouway/crm/core/Tree/HeterogeneousTree.java:
```diff
@@ -0,0 +1,145 @@
+     */
+    private HetTreeElement findHead(HetTreeElement node){
+
+        if(!node.isFull){
```
node.isFull ? I read it like node has limited number of elements as defined that way.

tsbonev commented:
Drew the car class diagram on a piece of paper
Removed the SumatorExceptions class and implemented its Exception in the Sumator class
Removed the success flag from Sumator and the finally statement
Removed the static int for elemCount, now each tree counts its own elements as an instance variable and passes it on to the elements so that they can properly keep their index
Removed unused treeSize integer

I use the node.isFull flag to ensure that the tree is binary, so that if a call is made to the tree to add an element, if the current head already has two children, the head moves to one of them and adds the element into the available spot

mgenov commented on src/com/clouway/crm/core/Sumator/ExceptionList.java:
```diff
@@ -0,0 +1,50 @@
+                throw new IndexOutOfBoundsException();
+            }
+            attributes.add(obj);
+            System.out.println("Successfully added " + obj);
```
System.outs should not be placed in the production code for now.

mgenov commented on src/com/clouway/crm/core/Sumator/ExceptionList.java:
```diff
@@ -0,0 +1,50 @@
+        this.attributes = new ArrayList<Object>(SIZE);
+    }
+
+    public void Add(Object obj){
```
methods in java should start with lower cases character

mgenov commented on src/com/clouway/crm/core/Sumator/ExceptionList.java:
```diff
@@ -0,0 +1,50 @@
+        }
+    }
+
+    public void Remove(){
```
.

mgenov commented on src/com/clouway/crm/core/Sumator/ExceptionList.java:
```diff
@@ -0,0 +1,50 @@
+            System.out.println("Successfully added " + obj);
+        }
+        catch (IndexOutOfBoundsException e){
+            System.out.println("Caught exception: " + e);
```
Method could return true or false if adding wasn't successfully executed.

mgenov commented on src/com/clouway/crm/core/Sumator/Sumator.java:
```diff
@@ -0,0 +1,55 @@
+            return res;
+        }
+        catch (NumberFormatException e){
+            return "Caught exception: " + e;
```
It is better exception to be re-thrown as user have to check for "Caught exception: " result which adds mess to the code.

mgenov commented on src/com/clouway/crm/core/Tree/HetTreeElement.java:
```diff
@@ -0,0 +1,30 @@
+            this.left = new HetTreeElement(object, elemCount);
+            return false;
+        }
+        else if(right == null){
```
Regarding isFull comment: If right or left are null then nothing is going to happen. Also keep in mind that logic first is looking for the node which is having an empty left or right node to be able to add the new node there.

mgenov replied:
> Regarding isFull: I can't think of another method to only ever give a tree element two child nodes besides having them coded into the class and checking if they are null

You are doing it and you have to do it as you have to check whether this is the end node to which the new one need to be added.

mgenov commented on src/com/clouway/crm/core/Tree/HeterogeneousTree.java:
```diff
@@ -0,0 +1,144 @@
+     */
+    public void printElements(String order){
+
+        if(order == "pre-order"){
```
it's not a good idea == to be used for equality check in java

```java
String a = "abc";
String b = "abc"; 

// inlined by the compiler and a and b are pointing to the same memory and this is why == succeeds 

String a = new String("abc")
String b = new String("abc")
a == b // returns false ! 
```

mgenov commented on src/com/clouway/crm/core/Tree/HeterogeneousTree.java:
```diff
@@ -0,0 +1,144 @@
+        else if(order == "in-order"){
+            //ArrayList<Integer> result = new ArrayList<>();
+            inOrder(parentNode);
+            //int[] sortedResult = result.stream().mapToInt(c -> c).toArray();
```
Commented code need to be removed

mgenov commented on src/com/clouway/crm/core/Tree/HomogeneousTree.java:
```diff
@@ -0,0 +1,187 @@
+        else if(order == "in-order"){
+            //ArrayList<Integer> result = new ArrayList<>();
+            inOrder(parentNode);
+            //int[] sortedResult = result.stream().mapToInt(c -> c).toArray();
```
Commented code need to be removed

mgenov commented:
The commits need to be squashed when you make your changes.

tsbonev commented:
Squashed commits
Made methods lowercase, removed system out from code, rethrew exceptions, returned true when successfully adding
Removed commented code
Regarding isFull: I can't think of another method to only ever give a tree element two child nodes besides having them coded into the class and checking if they are null
